scripts/lib/ts-generator.js

```diff
@@ -0,0 +1,437 @@
+        if (match) {
+            var model = match[1];
+            if (!foundModels[model]) {
+                foundModels[model] = modelFileName;
+            }
+        }
+    });
+
+    // now update the metadata
+    var keys = Object.keys(metadata);
+    keys.forEach(function(key) {
+        var modelName = metadata[key].prefix || key;
+        metadata[key].api_path = foundModels[modelName];
+        // find waiters file
+        var baseName = foundModels[modelName].split('.')[0];
+        if (modelPaths.indexOf(baseName + '.waiters2.json') >= 0) {
+            metadata[key].waiters_path = baseName + '.waiters2.json';
+        }
+    });
+};
+
+/**
+ * Generates the file containing DocumentClient interfaces.
+ */
+TSGenerator.prototype.generateDocumentClientInterfaces = function generateDocumentClientInterfaces() {
+    var self = this;
+    // get the dynamodb model
+    var dynamodbModel = this.loadServiceApi('dynamodb');
+    var code = '';
+    // include node reference
+    code += '///<reference types="node" />\n';
+    // generate shapes
+    var modelShapes = dynamodbModel.shapes;
+    // iterate over each shape
+    var shapeKeys = Object.keys(modelShapes);
+    shapeKeys.forEach(function (shapeKey) {
+        var modelShape = modelShapes[shapeKey];
+        // ignore exceptions
+        if (modelShape.exception) {
+            return;
+        }
+        // overwrite AttributeValue
+        if (shapeKey === 'AttributeValue') {
+            code += self.generateDocString('A JavaScript object or native type.');
+            code += 'export type ' + shapeKey + ' = any;\n';
+            return;
+        }
+        code += self.generateTypingsFromShape(shapeKey, modelShape);
+    });
+
+    // write file:
+    this.writeTypingsFile('document_client_interfaces', path.join(this._sdkRootDir, 'lib', 'dynamodb'), code);
+};
+
+/**
+ * Returns a service model based on the serviceIdentifier.
+ */
+TSGenerator.prototype.loadServiceApi = function loadServiceApi(serviceIdentifier) {
+    // first, find the correct identifier
+    var metadata = this.metadata;
+    var serviceFilePath = path.join(this._apiRootDir, metadata[serviceIdentifier].api_path);
+    var serviceModelFile = fs.readFileSync(serviceFilePath);
+    var serviceModel = JSON.parse(serviceModelFile);
+    // load waiters file if it exists
+    var waiterFilePath;
+    if (metadata[serviceIdentifier].waiters_path) {
+        waiterFilePath = path.join(this._apiRootDir, metadata[serviceIdentifier].waiters_path);
+        var waiterModelFile = fs.readFileSync(waiterFilePath);
+        var waiterModel = JSON.parse(waiterModelFile);
+        serviceModel.waiters = waiterModel.waiters;
+    }
+
+    return serviceModel;
+};
+
+/**
+ * Determines if a member is required by checking for it in a list.
+ */
+TSGenerator.prototype.checkRequired = function checkRequired(list, member) {
+    if (list.indexOf(member)) {
```

Comment:
Otherwise it's always returning true if the member isn't actually required, and it's returning false if the member happens to be the first one in the list